### ISSUE TITLE
Scheduler exception while handling exception in Linux+Python3

### ISSUE THREAD

**Issue by ASalazarMX** [web2py/web2py#1999]
The scheduler processes only one task task, the remaining tasks stay queued.

I ran the scheduler manually and submitted a simple task (adding two numbers) several times. After running the first task, the remaining tasks stay queued, and the worker's hertbeat stops updating in the database.

Running the scheduler as unprivileged user or as root makes no difference. Issue doesn't happen if the scheduler is run with Python 2 on the same configuration.
```
$ sudo python3 /var/www/web2py/web2py.py -K extranet
web2py Web Framework
Created by Massimo Di Pierro, Copyright 2007-2018
Version 2.17.1-stable+timestamp.2018.08.06.01.02.56
Database drivers available: sqlite3, psycopg2, imaplib, pymysql, pyodbc
starting single-scheduler for "extranet"...
ERROR:web2py.scheduler.vavextranetqa#7733:Error retrieving status
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/var/www/web2py/gluon/scheduler.py", line 1168, in send_heartbeat
    mybackedstatus = db(sw.worker_name == self.worker_name).select().first()
  File "/var/www/web2py/gluon/packages/dal/pydal/objects.py", line 2277, in select
    return adapter.select(self.query, fields, attributes)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 763, in select
    return self._select_aux(sql, fields, attributes, colnames)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 719, in _select_aux
    rows = self._select_aux_execute(sql)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 713, in _select_aux_execute
    self.execute(sql)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/__init__.py", line 67, in wrap
    return f(*args, **kwargs)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 413, in execute
    rv = self.cursor.execute(command, *args[1:], **kwargs)
psycopg2.OperationalError: SSL connection has been closed unexpectedly


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/var/www/web2py/gluon/scheduler.py", line 637, in run
    self.send_heartbeat(counter)
  File "/var/www/web2py/gluon/scheduler.py", line 1239, in send_heartbeat
    db.rollback()
  File "/var/www/web2py/gluon/packages/dal/pydal/base.py", line 708, in rollback
    self._adapter.rollback()
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/__init__.py", line 56, in wrap
    return f(*args, **kwargs)
  File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 866, in rollback
    return self.connection.rollback()
psycopg2.InterfaceError: connection already closed

DEBUG:web2py.app.extranet:    new task report: COMPLETED
```

Running fresh install of Ubuntu Linux 18.04.1 + PostgreSQL 10 + Python 3.

Database connection: db = DAL('postgres:psycopg2://user:password@127.0.0.1/extranet', pool_size=10)

**Comment by ASalazarMX** [comment]
Link to discussion of possible fix in web2py-users: https://groups.google.com/d/topic/web2py/VFMDlza07PI/discussion

**Comment by leonelcamara** [comment]
This was fixed, please close.

**Comment by ASalazarMX** [comment]
Tried again with the new 2.18.5-stable+timestamp.2019.04.08.04.22.03. OS is Ubuntu 18.04 LTS, Python is v3.6.8 and PostgreSQL is v10.9.

Still hangs after a couple of hors/days (variable). This is the log of the last executed task before the hang. The task checks a database queue and does very sporadic emails.

```
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:   work to do 6
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:    new scheduler_run record
Jul 27 18:27:37 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:new task 6 "task_mailer" appname/ti_scheduler.task_mailer
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315: new task allocated: appname/ti_scheduler.task_mailer
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:   task starting
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:    task started
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:........recording heartbeat (RUNNING)
Jul 27 18:27:37 servername sudo[19300]: ERROR:web2py.scheduler.servername#19315:Error retrieving status
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.app.appname:    new task report: COMPLETED
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.app.appname:   result: "[18:27:37 MAILER 0.00s] Se enviaron 0 de 0 correos pendientes"
Jul 27 18:27:37 servername sudo[19300]: Exception in thread Thread-1:
Jul 27 18:27:37 servername sudo[19300]: Traceback (most recent call last):
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/scheduler.py", line 1211, in send_heartbeat
Jul 27 18:27:37 servername sudo[19300]:     worker_stats=self.w_stats)
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/objects.py", line 2434, in update
Jul 27 18:27:37 servername sudo[19300]:     ret = db._adapter.update(table, self.query, row.op_values())
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 524, in update
Jul 27 18:27:37 servername sudo[19300]:     raise e
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 519, in update
Jul 27 18:27:37 servername sudo[19300]:     self.execute(sql)
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/__init__.py", line 67, in wrap
Jul 27 18:27:37 servername sudo[19300]:     return f(*args, **kwargs)
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 413, in execute
Jul 27 18:27:37 servername sudo[19300]:     rv = self.cursor.execute(command, *args[1:], **kwargs)
Jul 27 18:27:37 servername sudo[19300]: psycopg2.OperationalError: SSL error: sslv3 alert bad record mac
Jul 27 18:27:37 servername sudo[19300]: During handling of the above exception, another exception occurred:
Jul 27 18:27:37 servername sudo[19300]: Traceback (most recent call last):
Jul 27 18:27:37 servername sudo[19300]:   File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
Jul 27 18:27:37 servername sudo[19300]:     self.run()
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/scheduler.py", line 635, in run
Jul 27 18:27:37 servername sudo[19300]:     self.send_heartbeat(counter)
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/scheduler.py", line 1247, in send_heartbeat
Jul 27 18:27:37 servername sudo[19300]:     db.rollback()
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/base.py", line 710, in rollback
Jul 27 18:27:37 servername sudo[19300]:     self._adapter.rollback()
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/__init__.py", line 56, in wrap
Jul 27 18:27:37 servername sudo[19300]:     return f(*args, **kwargs)
Jul 27 18:27:37 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 866, in rollback
Jul 27 18:27:37 servername sudo[19300]:     return self.connection.rollback()
Jul 27 18:27:37 servername sudo[19300]: psycopg2.InterfaceError: connection already closed
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:  task completed or failed
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315: recording task report in db (COMPLETED)
Jul 27 18:27:37 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:task completed (COMPLETED)
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 27 18:27:37 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:nothing to do
Jul 27 18:27:37 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:sleeping...
Jul 27 18:27:40 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 27 18:27:40 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:nothing to do
Jul 27 18:27:40 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:sleeping...
Jul 27 18:27:43 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 27 18:27:43 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:nothing to do
```
From this point forward, the last two messages repeat and the workers remain unresponsive. 

PGAdmin shows a connection idle in transaction with the following query:
```
SELECT "scheduler_task"."id", "scheduler_task"."application_name", "scheduler_task"."task_name", "scheduler_task"."group_name", "scheduler_task"."status", "scheduler_task"."broadcast", "scheduler_task"."function_name", "scheduler_task"."uuid", "scheduler_task"."args", "scheduler_task"."vars", "scheduler_task"."enabled", "scheduler_task"."start_time", "scheduler_task"."next_run_time", "scheduler_task"."stop_time", "scheduler_task"."repeats", "scheduler_task"."retry_failed", "scheduler_task"."period", "scheduler_task"."prevent_drift", "scheduler_task"."cronline", "scheduler_task"."timeout", "scheduler_task"."sync_output", "scheduler_task"."times_run", "scheduler_task"."times_failed", "scheduler_task"."last_run_time", "scheduler_task"."assigned_worker_name" FROM "scheduler_task" WHERE (("scheduler_task"."assigned_worker_name" = 'servername#19315') AND ("scheduler_task"."status" = 'ASSIGNED')) ORDER BY "scheduler_task"."next_run_time" LIMIT 1 OFFSET 0;
```

Terminating this session seems to crash the scheduler, which gets restarted by systemd and resumes processing tasks:
```
Jul 29 10:34:56 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 29 10:34:56 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:nothing to do
Jul 29 10:34:56 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:sleeping...
Jul 29 10:34:59 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 29 10:34:59 servername sudo[19300]: INFO:web2py.scheduler.servername#19315:nothing to do
Jul 29 10:34:59 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:sleeping...
Jul 29 10:35:02 servername sudo[19300]: DEBUG:web2py.scheduler.servername#19315:looping...
Jul 29 10:35:02 servername sudo[19300]: Traceback (most recent call last):
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/shell.py", line 285, in run
Jul 29 10:35:02 servername sudo[19300]:     exec(python_code, _env)
Jul 29 10:35:02 servername sudo[19300]:   File "<string>", line 1, in <module>
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/scheduler.py", line 910, in loop
Jul 29 10:35:02 servername sudo[19300]:     task = self.wrapped_pop_task()
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/scheduler.py", line 963, in wrapped_pop_task
Jul 29 10:35:02 servername sudo[19300]:     db.commit()  # another nifty db.commit() only for Mysql
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/base.py", line 707, in commit
Jul 29 10:35:02 servername sudo[19300]:     self._adapter.commit()
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/__init__.py", line 56, in wrap
Jul 29 10:35:02 servername sudo[19300]:     return f(*args, **kwargs)
Jul 29 10:35:02 servername sudo[19300]:   File "/var/www/web2py/gluon/packages/dal/pydal/adapters/base.py", line 862, in commit
Jul 29 10:35:02 servername sudo[19300]:     return self.connection.commit()
Jul 29 10:35:02 servername sudo[19300]: psycopg2.OperationalError: terminating connection due to administrator command
Jul 29 10:35:02 servername sudo[19300]: server closed the connection unexpectedly
Jul 29 10:35:02 servername sudo[19300]:         This probably means the server terminated abnormally
Jul 29 10:35:02 servername sudo[19300]:         before or while processing the request.
Jul 29 10:35:03 servername sudo[19300]: pam_unix(sudo:session): session closed for user www-data
Jul 29 10:35:03 servername systemd[1]: web2py-scheduler.service: Service hold-off time over, scheduling restart.
Jul 29 10:35:03 servername systemd[1]: web2py-scheduler.service: Scheduled restart job, restart counter is at 1.
Jul 29 10:35:03 servername systemd[1]: Stopped Web2py task scheduler.
Jul 29 10:35:03 servername systemd[1]: Started Web2py task scheduler.
Jul 29 10:35:03 servername sudo[26840]:     root : TTY=unknown ; PWD=/ ; USER=www-data ; COMMAND=/usr/bin/python3 /var/www/web2py/web2py.py -D 0 -K appname
Jul 29 10:35:03 servername sudo[26840]: pam_unix(sudo:session): session opened for user www-data by (uid=0)
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:defining tables (migrate=True)
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:thread building own DAL object
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:04 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:defining tables (migrate=False)
Jul 29 10:35:04 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    freeing workers that have not sent heartbeat
Jul 29 10:35:04 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: I'm a ticker
Jul 29 10:35:07 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:07 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Assigning tasks...
Jul 29 10:35:07 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: workers are 1
Jul 29 10:35:07 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: tasks are 0
Jul 29 10:35:07 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Tasks assigned...
Jul 29 10:35:07 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:07 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:10 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:10 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:10 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:10 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:13 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:13 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:13 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:13 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:16 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:16 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:16 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:16 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:19 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:19 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:19 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:19 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:19 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    freeing workers that have not sent heartbeat
Jul 29 10:35:19 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: I'm a ticker
Jul 29 10:35:22 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:22 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Assigning tasks...
Jul 29 10:35:22 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: workers are 1
Jul 29 10:35:22 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: tasks are 3
Jul 29 10:35:22 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Tasks assigned...
Jul 29 10:35:22 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:22 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   work to do 6
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    new scheduler_run record
Jul 29 10:35:25 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:new task 6 "task_mailer" appname/ti_scheduler.task_mailer
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: new task allocated: appname/ti_scheduler.task_mailer
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   task starting
Jul 29 10:35:25 servername sudo[26840]: web2py Web Framework
Jul 29 10:35:25 servername sudo[26840]: Created by Massimo Di Pierro, Copyright 2007-2019
Jul 29 10:35:25 servername sudo[26840]: Version 2.18.5-stable+timestamp.2019.04.08.04.22.03
Jul 29 10:35:25 servername sudo[26840]: Database drivers available: sqlite3, psycopg2, imaplib, pymysql, pyodbc
Jul 29 10:35:25 servername sudo[26840]: starting single-scheduler for "appname"...
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    task started
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:thread building own DAL object
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:defining tables (migrate=False)
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (RUNNING)
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.app.appname:    new task report: COMPLETED
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.app.appname:   result: "[10:35:25 MAILER 0.00s] Se enviaron 0 de 0 correos pendientes"
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:  task completed or failed
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: recording task report in db (COMPLETED)
Jul 29 10:35:25 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:task completed (COMPLETED)
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   work to do 5
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    new scheduler_run record
Jul 29 10:35:25 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:new task 5 "task_appname_cierre" appname/ti_scheduler.task_appname_cierre
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: new task allocated: appname/ti_scheduler.task_appname_cierre
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   task starting
Jul 29 10:35:25 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    task started
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.app.appname:    new task report: COMPLETED
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.app.appname:   result: "[10:35:27 appname_CIERRE 1.66s] Cierre diario finalizado]"
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:  task completed or failed
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: recording task report in db (COMPLETED)
Jul 29 10:35:27 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:task completed (COMPLETED)
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   work to do 8
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    new scheduler_run record
Jul 29 10:35:27 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:new task 8 "task_housekeeping" appname/ti_scheduler.task_housekeeping
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: new task allocated: appname/ti_scheduler.task_housekeeping
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:   task starting
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    task started
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.app.appname:    new task report: COMPLETED
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.app.appname:   result: "[10:35:27 HOUSEKEEPING 0.01s] Limpieza diaria finalizada"
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:  task completed or failed
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852: recording task report in db (COMPLETED)
Jul 29 10:35:27 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:task completed (COMPLETED)
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:27 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:27 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:28 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:thread building own DAL object
Jul 29 10:35:28 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:defining tables (migrate=False)
Jul 29 10:35:28 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:30 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:30 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:30 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:31 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:33 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:33 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:nothing to do
Jul 29 10:35:33 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
Jul 29 10:35:34 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:........recording heartbeat (ACTIVE)
Jul 29 10:35:34 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:    freeing workers that have not sent heartbeat
Jul 29 10:35:34 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: I'm a ticker
Jul 29 10:35:36 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:looping...
Jul 29 10:35:36 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Assigning tasks...
Jul 29 10:35:36 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: workers are 1
Jul 29 10:35:36 servername sudo[26840]: INFO:web2py.scheduler.servername#26852:TICKER: tasks are 0
Jul 29 10:35:36 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:Tasks assigned...
Jul 29 10:35:36 servername sudo[26840]: DEBUG:web2py.scheduler.servername#26852:sleeping...
```

**Comment by ASalazarMX** [comment]
A temporary workaround for PostgreSQL 9.6+ is setting `idle_in_transaction_session_timeout` at the database level. This counts on systemd restarting the scheduler when it crashes:
```
ALTER DATABASE SET idle_in_transaction_session_timeout = '5min', 
```

**Comment by bmiklautz** [comment]
We were hit by this issue too (Ubuntu 18.04 - postgresql
10.10-0ubuntu0.18.04.1, python 3.6.8, web2py R-2.18.5, psycopg2 2.7.4-1).

After analyzing the problem in detail, I found out that this has to do with the fact that 'fork' is
used by the python multiprocessing package as default on Linux but psycopg2 (libpq) isn't
"multiprocess save".

What happens is that when the scheduler executes a Task, the scheduler process is forked and all open connections are copied to the child process. 
Once the Task is done and the process exits it closes all connections to the db that leaves the db connections in the scheduler process, which are the same, in an undefined state.
Commit 33c6dd965655fdb2468ff2c2b8efaf1bc9bc3014 addresses this problem but only fights the symptoms.

The quick fix, at least for python3, is to use the 'spawn' method from the multiprocessing package.
I've tested the following and run multiple schedulers with it for some days now:
```
diff --git a/web2py.py b/web2py.py
index ced8ff22..0ad1d407 100755
--- a/web2py.py
+++ b/web2py.py
@@ -3,7 +3,7 @@

 import os
 import sys
-from multiprocessing import freeze_support
+from multiprocessing import freeze_support, set_start_method
 # import gluon.import_all ##### This should be uncommented for py2exe.py

 if hasattr(sys, 'frozen'):
@@ -23,6 +23,7 @@ import gluon.widget
 # Start Web2py and Web2py cron service!
 if __name__ == '__main__':
     freeze_support()
+    set_start_method('spawn')
     if 'COVERAGE_PROCESS_START' in os.environ:
         try:
             import coverage
```
When I got time I'll create a pull request with an adapted version of this.

**Comment by ASalazarMX** [comment]
I run into this issue daily. I'll try the referenced fix, and close this issue if the scheduler remains stable after a week.

**Comment by ASalazarMX** [comment]
Cloned the repository and ran my app. I've attached the last part of journalctl for the scheduler. After it crashes no more messages are generated. If it ended instead of hanging, the service would restart automatically and I'd probably wouldn't notice.

[20191218_end.log](https://github.com/web2py/web2py/files/3979406/20191218_end.log)

**Comment by ASalazarMX** [comment]
Missed that to use the spawn method you have to change the instantiation of the scheduler in myapp/models/scheduler.py to:
`scheduler = Scheduler(db, use_spawn=True)`

After doing that, the scheduler crashes because of a typo in line 659 of gluon/scheduler.py:
`sel.process = p = ctx.Process(target=executor, args=(retq, task, outq))`
should be
`self.process = p = ctx.Process(target=executor, args=(retq, task, outq))`

Currently the scheduler is running fine. I'll close this issue if the scheduler remains stable for a week.

**Comment by timnyborg** [comment]
I've 

> Missed that to use the spawn method you have to change the instantiation of the scheduler in myapp/models/scheduler.py to:
> `scheduler = Scheduler(db, use_spawn=True)`
> 
> After doing that, the scheduler crashes because of a typo in line 659 of gluon/scheduler.py:
> `sel.process = p = ctx.Process(target=executor, args=(retq, task, outq))`
> should be
> `self.process = p = ctx.Process(target=executor, args=(retq, task, outq))`
> 
> Currently the scheduler is running fine. I'll close this issue if the scheduler remains stable for a week.

I've made a pull request to fix that typo.

Hopefully this will fix the issue I've run into since upgrading to Python 3, where pg8000 or psycopg2 will crash at 4am with a Broken Pipe or "psycopg2.OperationalError: server closed the connection unexpectedly"

**Comment by ASalazarMX** [comment]
I'm glad to say the scheduler is again stable. 7 days without problems.

If @timnyborg 's patch is merged, for my part, this issue can be closed.

**Comment by rostagnolisandro** [comment]
Hey there, I'm experiencing a problem related to this issue. I've posted a topic in the [community forum](https://groups.google.com/g/web2py/c/EvJv1qewIpw/m/u0nmCEEUBQAJ) but still I think I should comment it here. 

I've recently upgraded to web2py Version 2.24.1-stable+timestamp.2023.03.23.05.07.17
It's running on python 3.9.14, Rocky Linux RHEL9, using PostgreSQL 15.2 for database.
I'm running Scheduler(..., with_spawn=True) but all the tasks that are from a different app that the one from where the scheduler was run, fail with this traceback:

`Traceback (most recent call last): File "/home/limon/medios/gluon/scheduler.py", line 494, in executor functions = current._scheduler.tasks AttributeError: '_thread._local' object has no attribute '_scheduler'`

Let me explain a bit more: I have one web2py instance with several applications, each one with its own database. One of these apps is the main one and the scheduler connects to its database.
```
...web2py/applications/main/...
...web2py/applications/app1/...
...web2py/applications/app2/...
```

I run three scheduler workers for the main app with this command:
`/opt/virtualenvs/py39/bin/python /home/limon/web2py/web2py.py -K main,main,main`

In applications/main/models/scheduler.py I instantiate the Scheduler like this:
`scheduler = Scheduler(db, max_empty_runs=0, heartbeat=5, use_spawn=True)`

And then, from several parts of my application I queue tasks. Some of these tasks are defined in the "main" application, but some others are defined in "app1" or "app2". And since changing to Scheduler(..., use_spawn=True) all the tasks defined in "main" application run ok, but all the other ones fail with the traceback I showed before.
What am I missing?

**Comment by rostagnolisandro** [comment]
I think I've found the cause of the problem: when the scheduler runs in the environment of a specific application (notice in my case I run it with "web2py.py -K main") and a task is queued for a function that is defined in a different application, the scheduler fails to pick up the task at [this line](https://github.com/web2py/web2py/blob/b9a73402d81d3c7483decce4643c4fb29ea979f6/gluon/scheduler.py#LL491C17-L491C17) because the "current" object has not attribute "_scheduler" nor "_scheduler.tasks".

I've made a simple fix that works ok and I think it doesn't break anything. I'll post a separated issue and I'll make a pull request, hope it helps.
